### PR TITLE
Assortativity

### PR DESCRIPTION
**Pull request by jpfairbanks**
This PR is a duplicate of #274 covering issues #273 and #273.

We add functions for computing assortativity with respect to degree and categorical labels with nominal and local variations.

### CODE REVIEW

**Comment by sbromberger** [comment]
Is the plan to fix the tests?

File: src/centrality/assortativity.jl [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+end
+
+"""
+(degree) assortativity - directed: 
+
+see equation (21) in M. E. J. Newman: Mixing patterns in networks, Phys. Rev. E 67, 026126 (2003), 
+http://arxiv.org/abs/cond-mat/0209450
+"""
+function assortativity_coefficient(g::DiGraph, sjk, ssj, sk, sjs, sks, nue)
+    return res = (sjk - sj*sk/nue)/sqrt((sjs - sj^2/nue)*(sks - sk^2/nue))
```
Comment (sbromberger) [inline src/centrality/assortativity.jl]: These parameters need types.

File: src/centrality/assortativity.jl [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+end
+
+"""
+graph degree assortativity:
+
+see equation (2) in M. E. J. Newman: Mixing patterns in networks, Phys. Rev. E 67, 026126 (2003), 
+http://arxiv.org/abs/cond-mat/0209450
+"""
+function assortativity_degree(g)
```
Comment (sbromberger) [inline src/centrality/assortativity.jl]: type here please.

**Comment by jpfairbanks** [comment]
We shouldn't merge it until the math errors are resolved. @akarpf have you resolved the math errors?

**Comment by sbromberger** [comment]
It needs a more in-depth review as well... I've just skimmed so far.

**Comment by akarpf** [comment]
@sbromberger Sorry, I was extremely busy recently as I am currently commuting between two different cities. @jpfairbanks  Can you hint me to the maths errors? I was actually double-checking it with the igraph functions. Trying to do all that on the weekend.

**Comment by jpfairbanks** [comment]
If you look at this file, https://github.com/JuliaGraphs/LightGraphs.jl/blob/assortativity/test/centrality/assortativity.jl

it tests that if you look at the path graph on `n` vertices, the assortativity should be increasing in `n`.
Also the assortativity of a complete graph should be positive since all vertices have the same degree. If these are not true properties of assortativity, let me know. These are intuitive properties to me, but I didn't prove them from the definition.

Can create a list of graphs that you can test with both igraph and LightGraphs?
If you run the igraph function on them and get the values, we can include those values in the tests.

**Comment by akarpf** [comment]
@jpfairbanks In the graph version of the assortativity measure a value of >0 points to assortative mixing (+1 perfect assortativity), while a value < 0 points to disassortative mixing (-1 perfect disassortativty). Of the 0 the graph is non-assortative. The individual level scores are not constrained to be inside [-1,1] but should in sum yield the graph wide measure. 
This is the check I made, having compared the results for graph level assortativity with those from igraph before.

**Comment by jpfairbanks** [comment]
Where is the code that checks against the values produced by igraph? That code should be the basis of the necessary unit tests.

**Comment by akarpf** [comment]
@jpfairbanks Clear, this is what I going to deliver as soon as I find the time. However, once can check for the consistence of the code by summing up the node level assortativity values of an arbitrary graph. This sum should correspond to the graph level result. If not, then I uploaded the wrong code. I am ging to check this.

**Comment by sbromberger** [comment]
Hi,

What's the plan for this PR?

**Comment by jpfairbanks** [comment]
Hey @akarpf, did you find the time to fix things?

**Comment by akarpf** [comment]
@akarpf I more or less fixed it, I however asked you a question with regard to example networks (florentine marriage network etc.) with which I/we can benchmark other libraries like igraph around three weeks ago.

**Comment by jpfairbanks** [comment]
Oh I don't remember that. What was the question? There are example networks in generators.jl and datasets.jl. Generators are graphs that are parameterized usually by a size. and datasets are examples found in the literature.

**Comment by sbromberger** [comment]
@jpfairbanks @akarpf - what's the status of this PR?

**Comment by sbromberger** [comment]
@akarpf @jpfairbanks we're coming up on 8 months since the PR was originally opened, and 4 months since the last substantive comment. What's the plan?

**Comment by sbromberger** [comment]
Closing out due to inactivity. @jpfairbanks - feel free to reopen if you want to keep working on it.

**Pull request closed without merging**